File: src/main/java/api/Blacklist.java
package api;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import tool.Response;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Eldath on 2017/1/30 0030.
 *
 * @author Eldath
 */
public class Blacklist implements API {
    private static Logger logger = LoggerFactory.getLogger(Blacklist.class);
    private static List<String> allowList = new ArrayList<>();
    private static Blacklist instance = null;

    static Blacklist getInstance() {
        if (instance == null) instance = new Blacklist();
        return instance;
    }

    static {
        // CUSTOM 以下为允许操作黑名单的QQ号
        allowList.add("1464443139");
    }

    @Override
    public void doPost(JSONObject object) {
        String sender_uid = object.get("sender_uid").toString();
        String group_uid = object.get("group_uid").toString();
        String content = object.get("content").toString();
        String sender = object.get("sender").toString();
        String[] split;
        String action, toBan;
        if (!content.contains(" ")) {
            Response.responseGroup(group_uid, "@" + sender + " 您的指示格式不对辣！（｀Δ´）！");
            return;
        }
        split = content.split(" ");
        if (split.length < 3) {
            Response.responseGroup(group_uid, "@" + sender + " 您的指示格式不对辣！（｀Δ´）！");
            return;
        }
        action = split[2];
        toBan = split[3];
        for (String thisAllowUid : allowList)
            if (thisAllowUid.equals(sender_uid)) {
                if ("add".equals(action)) {
                    Response.responseGroup(group_uid, "@" + sender + " 帐号" + toBan + "现已被 阿瓦隆回答我 功能屏蔽。");
                    logger.info("Account " + toBan + " is baned by " + sender_uid + " : " + sender + ".");
                    XiaoIce.blackList.put(toBan, 4);
                    return;
                } else if ("remove".equals(action)) {
                    if (!XiaoIce.blackList.containsKey(sender_uid)) {
                        Response.responseGroup(group_uid, "@" + sender + " 好像帐号" + toBan + "没有被屏蔽过呢-。-");
                        return;
                    }
                    Response.responseGroup(group_uid, "@" + sender + " 帐号" + toBan + "的屏蔽已被解除(^.^)");
                    logger.info("Account " + toBan + " is allowed again by " + sender_uid + " : " + sender + ".");
                    XiaoIce.blackList.put(toBan, 0);
                    return;
                } else {
                    Response.responseGroup(group_uid, "@" + sender + " 您的指示格式不对辣！（｀Δ´）！");
                    return;
                }
            }
        Response.responseGroup(group_uid, "@" + sender + " 您没有权限辣！（｀Δ´）！");
    }
}

File: src/main/kotlin/avalon/group/Blacklist.kt
package avalon.group

import avalon.api.Flag.at
import avalon.api.getAllowArray
import avalon.tool.pool.Constants
import avalon.tool.pool.Constants.Basic.CURRENT_SERVLET
import avalon.tool.pool.Constants.Basic.LANG
import avalon.util.GroupConfig
import avalon.util.GroupMessage
import org.slf4j.LoggerFactory
import java.util.regex.Pattern

/**
 * Created by Eldath on 2017/10/5 0030.
 *
 * @author Eldath
 */
object Blacklist : GroupMessageResponder() {
	private val logger = LoggerFactory.getLogger(Blacklist.javaClass)

	override fun doPost(message: GroupMessage, groupConfig: GroupConfig) {
		val senderUid = message.senderUid
		val max = GroupMessageHandler.punishFrequency
		val content = message.content
		val sender = message.senderNickName
		val groupUid = message.groupUid
		val allowList = getAllowArray(groupConfig, "Blacklist_basic")
		val split = content.split(" ")
		val toBan = if (content.contains(Regex("add|remove"))) split[3].toLong() else 0
		val map = GroupMessageHandler.blacklistPeopleMap
		allowList!!.first { it == senderUid }.apply {
			when (split[2]) {
				"list" -> {
					if (Constants.Basic.DEBUG)
						message.response(map.keys.joinToString())
					if (map.isEmpty())
						message.response("${at(message)} ${LANG.getString("group.blacklist.empty_banned_list")}")
					else {
						val to = map.keys.joinToString { CURRENT_SERVLET.getGroupSenderNickname(groupUid, it) }
						message.response("${at(message)} ${LANG.getString("group.blacklist.banned_list_has").format(to)}")
					}
				}
				"add" -> {
					message.response("${at(message)} ${String.format(LANG.getString("group.blacklist.account_now_banned"), toBan)}")
					Blacklist.logger.info("account $toBan is baned by $senderUid : $sender.")
					GroupMessageHandler.addBlackListPeople(toBan, max)
				}
				"remove" -> {
					if (!map.containsKey(toBan))
						message.response("${at(message)} ${String.format(LANG.getString("group.blacklist.account_not_banned"), toBan)}")
					else {
						message.response("${at(message)} ${String.format(LANG.getString("group.blacklist.account_now_released"), toBan)}")
						Blacklist.logger.info("account $toBan is allowed again by $senderUid : $sender.")
						GroupMessageHandler.addBlackListPeople(toBan, 0)
					}
				}
				else -> message.response("${at(message)} ${LANG.getString("group.blacklist.incorrect")}")
			}
		}
	}

	override fun responderInfo(): ResponderInfo = ResponderInfo(
			Pair("blacklist (list|add|remove)", LANG.getString("group.blacklist.help")),
			Pattern.compile("blacklist (list|add|remove)"),
			configIdentifier = arrayOf("Blacklist_basic"),
			manageable = false,
			permission = ResponderPermission.ADMIN)

	override fun instance() = this
}